Add `experimental.mace` architecture

pfebrer opened the pull request:
This PR adds the MACE architecture in metatrain. It replaces #832 to start with a good description highlighting the details of the implementation and problems/choices for which I would like people's opinions.

The general idea is to reuse as much as possible from MACE's own training scripts so that (1) it is more maintainable and easily kept in sync, (2) we just trust that the MACE people know what the good default hyperparameters are. 

# The implementation

## Model

For the model, we just import the MACE model and wrap it with:
- Data preprocessing (`Scaler` and `CompositionModel`).
- Heads that take the node features from MACE and produce the target properties with a linear layer.

## Trainer

For the trainer, it is the same trainer as PET, except for:
- The removal of rotational augmentation.
- The use of the same optimizers and learning rate schedulers that are used in MACE's own training scripts.
- This last point requires a minor difference in the training loop, which is to pass the loss to `lr_scheduler.step(metrics=loss)`.

The rest is just the same as PET, and I think it could be beneficial to have code that can be shared, to have a consistent way in which "`metatrain` does things", but this is probably for another (potentially never-ending) discussion :)

## How to test it

`pip install metatrain[mace]`, then `pip install e3nn>0.5.0` (see the problems section)

The config file needs:

```yaml
architecture:
  name: experimental.mace  
  model:
    hidden_irreps: 64x0e + 64x1o + 64x2e
```

and then the rest of details for your dataset. There is no default for the hidden irreps in MACE's training CLI, this is why we need to pass it.

# Problems

1. **We use linear layers for all heads, but the way MACE really predicts invariant targets is with an MLP**. One can gather all invariant targets and make MACE output them in its internal MLP. This would require a bit of index keeping if there are multiple invariant targets, which wouldn't be beautiful. The other option is to just create our own external MLPs, one for each invariant target.
**PR merged with the following solution:** *The arbitrary heads that we add are also MLPs. One can also choose to use MACE's internal head directly for a target with the ``mace_head_target`` parameter.*

2. **If we wanted to use MACE's internal MLP, it is not torchscript compatible**. Or at least this is what I get when metatrain wants to export the model. This can be tested with:
    ```yaml
    architecture:
      name: experimental.mace  
      model:
         hidden_irreps: 64x0e + 64x1o + 64x2e
         gate: silu # (The default in metatrain for now is None, which means no activation function I think)
    ```
    if some torchscript magician can fix it, that would be great.
    **PR merged with the following solution:** *``e3nn`` had a specific function ``e3nn.util.jit.compile`` which is the one to use for converting to torchscript. We use it and there is no problem.*

3. **MACE scripts don't really use `MACE`, but `ScaleShiftMACE`**. This version of mace accepts a scale and a shift on initialization and then applies it on forward, but I'm unsure if/how it handles multiple targets or targets with $\ell > 0$. For this reason, I decided for now to just simply use the scaler/composition model infrastructure in metatrain in combination with the bare `MACE` model. However, this might create problems when wanting to finetune foundational MACE models, which are (I think) `ScaleShiftMACE`. I thought about having a `MetaMACE.from_mace` classmethod that takes care of checking these things and probably avoids using the scaler and composition models, kind of like a `restart`.
**PR merged with the following solution:** *If a MACE model is loaded, we move the scales and shifts from MACE's internal modules to metatrain's `CompositionModel` and `Scaler`, for the target defined in the ``mace_head_target`` hyper*

4. **Discussion needed on what goes to metatomic and what goes to metatrain**. In previous discussions, it was mentioned that it would be nice to have the simple functionality of wrapping a MACE model to make it compatible with metatomic. There is clearly a lot of overlap between this and the current definition of the `MetaMACE` model, so we need to think on how to distribute the code.
**PR merged with the following solution:** *For now, everything goes to metatrain. A model is exported from MACE file to metatomic by training with 0 epochs. This is documented in the architecture's page.*

5. **Batching efficiency is probably far from optimal**, and might be a bottleneck, have not checked. Since I just let metatrain handle the data:
    - The metatrain batch is created.
    - The batch is unpacked into a list of systems.
    - The list of systems is converted into a `torch_geometric` batch, which is what MACE takes as input.
    This is all done inside the `forward` method. I don't know what is the best way to solve this, some solutions might interfere a lot with the way metatrain handles data, while one could also think of a function (metatrain batch --> torch_geometric batch) that does the conversion without unpacking. 
**PR merged with the following solution:** *We still create the batch in ``forward``, since it doesn't seem to be a bottleneck for now*

6. **MACE pins a version of `e3nn` that is not torchscript compatible**. The reasons why this pin is still there are unclear to me, but you can read the discussion in https://github.com/ACEsuit/mace/issues/555. Anyway, the reason why it's not torchscript compatible is quite stupid: there's a missing `int` type hint in a function argument. Newer versions of e3nn work fine with MACE, so one can install a new version of e3nn after installing `metatrain[mace]`. However, this is again probably going to interfere with loading MACE foundational models (which use e3nn==0.4.4 I presume). Something that we could do, maybe, is to type hint the problematic function at runtime. Unless torchscript uses the source code, and not the runtime python object.
**PR merged with the following solution:** *With the special ``e3nn`` compiling function (see solution to problem 2), converting ``e3nn==0.4.4`` to torchscript works, so no need for anything to be done. We don't specify a version for ``e3nn`` so MACE's pin is effectively used for now.*

Looking forward to your opinions to try to shape this into the best implementation possible :)

------------------------------------------------

# Contributor (creator of pull-request) checklist

- [x] Add your architecture to the `experimental` or `stable` folder. See the
  [docs/src/dev-docs/architecture-life-cycle.rst](Architecture life cycle)
  document for requirements. `src/metatrain/experimental/<architecture_name>`
- [x] Add default hyperparameter file to
  `src/metatrain/experimental/<architecture_name>/default-hypers.yml`
- [x] Add your architecture to the CI in `.github/workflow/architecture-tests.yml`
- [x] Add a new dependencies entry in the `optional-dependencies` section in the
  `pyproject.toml`
- [ ] Add tests:
  - [x] checking that the code is compatible with TorchScript
  - [ ] checking the basic functionality (invariance, fitting, prediction)
  - [ ] checking that the checkpoints are properly versionned (see the existing
    `test_checkpoint.py` in other architectures)
- [x] Add maintainers as codeowners in [CODEOWNERS](CODEOWNERS)
- [ ] Trigger a GPU test by asking a maintainer to comment "cscs-ci run".

# Reviewer checklist

## New experimental architectures

- [ ] Capability to fit at least a single quantity and predict it, verified through CI
   tests.
- [x] Compatibility with JIT compilation using `TorchScript
   <https://pytorch.org/docs/stable/jit.html>`_.
- [x] Provision of reasonable default hyperparameters.
- [x] A contact person designated as the maintainer, mentioned in `__maintainers__` and the `CODEOWNERS` file
- [x] All external dependencies must be pip-installable. While not required to be on
   PyPI, a public git repository or another public URL with a repository is acceptable.

<!-- readthedocs-preview metatrain start -->
----
📚 Documentation preview 📚: https://metatrain--846.org.readthedocs.build/en/846/

<!-- readthedocs-preview metatrain end -->

pfebrer commented on src/metatrain/experimental/mace/tests/__init__.py:
```diff
@@ -0,0 +1,12 @@
+    Path(__file__).parents[4] / "tests/resources/carbon_reduced_100.xyz"
+)
+
+DEFAULT_HYPERS = get_default_hypers("pet")
```
I think it is better to get mace's hypers :sweat_smile:

PicoCentauri commented:
I just want to bring up that the required paramter for mace

```
architecture:
  name: experimental.mace  
  model:
    hidden_irreps: 64x0e + 64x1o + 64x2e
```

is kinda against our idea of having every model with defaults. I discussed with @pfebrer and it might be hard to define some default size that runs on many systems. I am fine with doing this but just want some comments from the others. @frostedoyster and @Luthaf

frostedoyster commented:
Thanks everyone and especially @pfebrer!

Here is what I would do: in short, I would disregard loading the exact foundation models from MACE. In my experience, fine-tuning is insensitive to changes in composition and scaling, so I would take a `ScaleShiftMACE` (i.e., MACE with composition model and scaler), extract its MACE, and use metatrain's composition and scaler for the fine-tuning run. This will IMO give us equal or better fine-tuning compared to the MACE repo, since our scaler technology is quite advanced, and allow fine-tuning of pre-trained MACE models on any target we want. Yes, we would lose the ability of re-exporting existing foundation models, but 1) people can already use MACE foundation models and I doubt they would be interested in re-wrapping it in our stuff for the time being (until we offer GROMACS), and 2) metatrain is about training and not wrapping models (in this sense, a metatomic wrapper of MACE models might belong somewhere else). What we can offer here is adding the flexibility to train on pretty much arbitrary targets using MACE without writing code, and including starting training from their pre-trained models.

Therefore, here are my suggestions:

1. We use our MLP (see heads in other architectures) and copy the weights from MACE when fine-tuning if possible and not too cumbersome, otherwise re-initialization is ok
2. See point 1
3. We use metatrain's composition and scaler
4. Metatrain (CLI) can offer fine-tuning; exact wrapping should be separate IMO. I'm not opposed to `from metatrain.mace import wrap_as_metatomic`, if we don't find a better place for it. People would have to call it from Python and then `torch.jit.save()` the resulting model
5. I will check this before we merge and then we can optimize it if it's a bottleneck
6. I would try to avoid a manual e3nn installation outside `pip install metatrain[mace]`. I would definitely patch the missing `int()` upon importing `metatrain.experimental.mace`.

For the missing default hypers: I'm not a fan, but we can error out if the user doesn't provide it. Effectively the MACE default used to be `128x0e + 128x1o`, so I would include that as a baseline suggestion in the error message. Another alternative is to make that the default.

I'd be happy to help with the missing tests.

PicoCentauri commented:
Does the standard lib offer such thing. At least I am not aware. 

I think the patch is fine and @pfebrer maybe do a PR upstream on e3nn to change it.

pfebrer commented:
Ok, I think until somebody is interested enough that is willing to test things it is better to not add the capability of finetuning an original MACE model. I don't think it makes sense to implement a new way of doing things without having a practical use or at least good testing proving that it is the right thing to do.

So I would say for now in this PR we stick to supporting only training and finetuning models that have been trained in metatrain, then if somebody has a different use case we tell them to submit an issue and adapt the implementation.

pfebrer commented on src/metatrain/utils/data/target_info.py:
```diff
@@ -166,7 +166,10 @@ def _check_layout(self, layout: TensorMap) -> None:
+            if len(layout.keys.names) < 2 or layout.keys.names[:2] != [
+                "o3_lambda",
+                "o3_sigma",
+            ]:
```
@frostedoyster do you think this will break something? Essentially the concatenated features of MACE have irreps that look like:

```python
o3.Irreps("2x0e + 2x1o + 2x0e")
```

(i.e. they have two scalar parts that are separated). I could reorder them so that I have `"4x0e + 2x1o"` but the most natural thing I believe is to keep their order. To convert those features into a `TensorMap` I need that blocks have some extra key that can differentiate the first scalar block from the last one, so I have keys `["o3_lambda", "o3_sigma", "i_irrep"]`.

pfebrer replied:
Can properties be non-unique? I.e. `i_irrep=[0,0,0,1,1]`?

pfebrer commented:
cscs-ci run

pfebrer commented:
cscs-ci run

pfebrer commented:
Ready for review :+1:

Thanks to Guillaume we found some bugs, and now I added a test to make sure that a MACE foundation model produces exactly the same energies and forces when run natively vs when wrapped in metatomic. So quite confident now that the implementation is good :)

PicoCentauri commented on .github/workflows/architecture-tests.yml:
```diff
@@ -19,6 +19,7 @@ jobs:
           - pet
           - nanopet
           - flashmd
+          - mace
```
Just for my personal zen.

Should we sort alphabetically?

PicoCentauri commented on src/metatrain/experimental/mace/_dev/documentation.py_template:
maybe make this a py file to have syntax highlight and linting? Or will the linter complain?

PicoCentauri commented on src/metatrain/experimental/mace/modules/heads.py:
```diff
@@ -0,0 +1,162 @@
+                    for ir in irreps_in
+                    if ir.ir in output_irreps
+                ],
+                o3.Irreps(""),
```
How does `sum` work on a list of strings?

PicoCentauri commented on src/metatrain/experimental/mace/modules/heads.py:
```diff
@@ -0,0 +1,162 @@
+        last_layer_features_irreps = o3.Irreps(self.hidden_irreps)
+        for output_ir in output_irreps:
+            if last_layer_features_irreps.count(output_ir) == 0:
+                last_layer_features_irreps += 0 * output_ir
```
Isn't adding zeros unnecessary. Why can't you skip this operation?

pfebrer replied:
It is explained in the comment above the code :)

PicoCentauri commented on src/metatrain/experimental/mace/modules/scale_shift.py:
```diff
@@ -0,0 +1,11 @@
+import torch
+
+
+class FakeScaleShift(torch.nn.Module):
```
I agree with @copilot here. A docstring in a public class would be nice.

PicoCentauri commented on src/metatrain/experimental/mace/tests/__init__.py:
```diff
@@ -0,0 +1,9 @@
+from pathlib import Path
+
+
+mace_model_url = "https://github.com/ACEsuit/mace-mp/releases/download/mace_mp_0/2023-12-10-mace-128-L0_energy_epoch-249.model"
```
A 2 year old model?

PicoCentauri commented on src/metatrain/experimental/mace/utils/hypers.py:
```diff
@@ -0,0 +1,100 @@
+
+
+# Keys from the MACE argparser that correspond to model hyperparameters
+MACE_MODEL_ARG_KEYS = [
```
Do we have a test that these are valid, if they release a new version?

PicoCentauri commented on src/metatrain/experimental/mace/utils/hypers.py:
```diff
@@ -0,0 +1,100 @@
+
+    def _get_type(action: argparse.Action) -> str:
+        if action.dest == "radial_MLP":
+            assert action.default == "[64, 64, 64]"
```
Is this real validation or just something for debdugging. If the user might see this error we should transform it into a real error message.

PicoCentauri commented on src/metatrain/experimental/mace/documentation.py:
```diff
@@ -0,0 +1,203 @@
+"""
```
I would ad a comment in the first line saying that this is a generated file. Do not edit. Rather edit the source file at XXX

pfebrer replied:
Actually I don't know what to do with this :sweat_smile: Right now the generating script just generates a documentation file so that you can take whatever you want from it, it doesn't replace the real `documentation.py`. This is because I think it could make sense to document better the hypers. MACE's documentation strings are really poor...

PicoCentauri commented on src/metatrain/experimental/mace/model.py:
```diff
@@ -0,0 +1,658 @@
+
+        self.finetune_config: Dict[str, Any] = {}
+
+    def restart(self, dataset_info: DatasetInfo) -> "MetaMACE":
```
```suggestion
    def restart(self, dataset_info: DatasetInfo) -> Self:
```

or is this forbidden by torch script.

pfebrer replied:
Is metatrain python >= 3.11?

PicoCentauri commented on tests/resources/periodic_water.lmp:
this is a data file. We should also put a `.data` suffix which I think is the more canonical name.

PicoCentauri reviewed (dismissed):
Amazing work @pfebrer  and @jwa7

pfebrer commented:
Amazing, you really can understand humans

pfebrer commented on README.md:
```diff
@@ -41,6 +41,7 @@ model:
 | NanoPET *(deprecated)*   | Re-implementation of the original PET with slightly improved training and evaluation speed                                           |
 | SOAP BPNN                | A Behler-Parrinello neural network with SOAP features                                                                                |
 | FlashMD                  | An architecture for the direct prediction of molecular dynamics                                                                      |
+| MACE                     | A higher order equivariant message passing neural network.
```
Done

pfebrer commented on src/metatrain/experimental/mace/_dev/documentation.py_template:
```diff
@@ -0,0 +1,156 @@
+
+This architecture is a very thin wrapper around the official
+MACE implementation, which is
+`hosted here <https://github.com/ACEsuit/mace>`_. Arbitrary heads
```
I've been trying to look at some description by the MACE people but couldn't find any. i have just added a one line description

pfebrer commented on src/metatrain/experimental/mace/utils/hypers.py:
```diff
@@ -0,0 +1,100 @@
+            optional = True
+
+        if isinstance(action, (argparse._StoreTrueAction, argparse._StoreFalseAction)):
+            annotation = "bool"
```
We just want to write these to a file as text. (this has been moved to `_gen_documentation.py`)

ceriottm commented on src/metatrain/experimental/mace/utils/mts.py:
Here the general docstring could guide a bit more.

ceriottm reviewed (commented):
General comments. Looks very good on the surface, thanks for the thanksless job!

pfebrer commented on src/metatrain/utils/testing/torchscript.py:
I think this should be done in a separate PR, since it is a documentation change to the common tests.

pfebrer commented:
Thanks @ceriottm! 

I addressed the suggestions and also now that I understand how the fixed weights in the scaler and composition model work I was able to transfer the baselines and scales from MACE to metatrain when a model file is loaded. Therefore @Luthaf "exporting" a foundation model now looks simpler:

`mtt train`
```yaml
architecture:
    name: experimental.mace
    model:
        mace_model: path/to/foundation/mace/model.model
        mace_head_target: energy
    training:
        num_epochs: 0
        batch_size: 1

training_set: dummy_dataset.xyz
validation_set: dummy_dataset.xyz
```

and I added also a programatic way (not publicly exposed) that could be useful for you to do experiments maybe:

```python
from metatrain.experimental.mace.utils._load_model_file import load_mace_model_file

model = load_mace_model_file(
    "path/to/foundation/mace/model.model",
    mace_head_target="energy",
    device=device,
)
```

no more changes now except to address reviews :sweat_smile:

pfebrer commented:
cscs-ci run

Luthaf commented on src/metatrain/experimental/mace/documentation.py:
```diff
@@ -0,0 +1,390 @@
+    tensor with higher order), the effective number of message passing steps for
+    that target will be ``num_interactions - 1``.
+    """
+    hidden_irreps: Optional[str] = "128x0e + 128x1o + 128x2e"
```
I don't understand why this is `Optional` when it has a default string?

Luthaf commented on src/metatrain/experimental/mace/documentation.py:
```diff
@@ -0,0 +1,390 @@
+    ] = "RealAgnosticResidualInteractionBlock"
+    """Name of interaction block.
+
+    Class that will be used to compute interactions between atoms at each layer.
```
should we link to the doc for these class in MACE doc?

pfebrer commented on src/metatrain/experimental/mace/_dev/_gen_documentation.py:
```diff
@@ -0,0 +1,62 @@
+
+
+if __name__ == "__main__":
+    write_mace_hypers_spec()
```
I prefer to keep the name a bit meaningful taking advantage that we are not in C if you don't mind :)

Luthaf commented on src/metatrain/experimental/mace/model.py:
```diff
@@ -0,0 +1,739 @@
+        }
+    )
+
+    # Attributes of the model. We can't uncomment these descriptions because
```
maybe you can remove the docstring duplicates? And only leave the type annotations

Luthaf replied:
my bad, I read them too fast and though they were hypers. all good!

Luthaf reviewed (commented):
Looks great, I mainly have a couple of questions about how the code works!

pfebrer commented on src/metatrain/experimental/mace/__init__.py:
```diff
@@ -0,0 +1,15 @@
+]
+
+__maintainers__ = [
+    ("Pol Febrer <pol.febrer@epfl.ch>", "@pfebrer"),
```
I don't know, he was the one that wrote this haha however he wants

pfebrer commented on src/metatrain/experimental/mace/documentation.py:
```diff
@@ -0,0 +1,390 @@
+        "RealAgnosticDensityResidualInteractionBlock",
+        "RealAgnosticResidualNonLinearInteractionBlock",
+    ] = "RealAgnosticResidualInteractionBlock"
+    """Name of interaction block for the first interaction layer.
```
Same, I don't think there's such a thing

pfebrer commented on src/metatrain/utils/testing/output.py:
```diff
@@ -66,7 +66,7 @@ def n_features(self) -> Optional[int]:
 
     @pytest.fixture
-    def n_last_layer_features(self) -> Optional[int]:
+    def n_last_layer_features(self) -> Optional[int | list[int]]:
```
Indeed, for now we only check energy last layer features so it's only an int, it's a leftover from trying to do something more fancy.

Reverted

pfebrer commented on src/metatrain/experimental/mace/tests/test_basic.py:
```diff
@@ -0,0 +1,165 @@
+    # the "features" output can have a well defined equivariant
+    # behavior, and therefore checking consistency when returning
+    # this output raises an error.
+    avoid_consistency_check = ["features"]
```
Done

Luthaf commented on src/metatrain/experimental/mace/documentation.py:
```diff
@@ -0,0 +1,390 @@
+    """Target to which the MACE head is related.
+
+    ``metatrain`` adds arbitrary heads on top of MACE to predict arbitrary targets.
+    However, MACE models have themselves a head. This hyperparameter specifies which
```
All of the published universal models are single head?

Luthaf reviewed (approved):
Looks good, thanks for the work @pfebrer!

Luthaf commented:
cscs-ci run
